### ISSUE TITLE
Unable to replicate DenseTNT results

### ISSUE THREAD

**Issue by rusty-electron** [alfredgu001324/MapUncertaintyPrediction#44]
Hi, 
As the title suggests I attempted to run the entire pipeline of your paper. So far I have replicated the experiments for MapTR and MapTRv2 and achieved the reported results in the paper (with small variance). Unfortunately, this is not the case for the trajectory prediction with DenseTNT 😟. I have reported one such result below:

### DenseTNT with MapTR - with uncertainty
<html xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=PowerPoint.Slide>
<meta name=Generator content="Microsoft PowerPoint 15">

</head>

<body>

Metric | Experiment | Reported
-- | -- | --
minADE | 1.2198 | 1.089
minFDE | 2.3284 | 2.006
MR | 0.4096 | 0.3499

</body>

</html>

This could ofc be an issue on my end so I want to roughly describe what I did. For the training script, I followed some conversations here in the issues and referred to your [other repo](https://github.com/alfredgu001324/MapBEVPrediction). I also copied the training hyperparameters from the paper appendix.

```bash
other code here
...
case $map_model in
  maptr)
    batch=64
    lr=0.0005
    dropout=0.5
    train_dir="../trj_data/maptr/train/data/"
    val_dir="../trj_data/maptr/val/data/"
    ;;
  maptrv2)
    batch=64
    lr=0.0005
    dropout=0.5
    train_dir="../trj_data/maptrv2/train/data/"
    val_dir="../trj_data/maptrv2/val/data/"
    ;;
  maptrv2-cl)
    batch=64
    lr=0.00018
    dropout=0.5
    train_dir="../trj_data/maptrv2/train/data/"
    val_dir="../trj_data/maptrv2/val/data/"
    ;;
  streammapnet)
    batch=64
    lr=0.001
    dropout=0.5
    train_dir="../trj_data/streammapnet/train/data/"
    val_dir="../trj_data/streammapnet/val/data/"
    ;;
  *)
    echo "Unknown map_model: $map_model"
    exit 1
    ;;
esac
```
#### training
```bash
  python src/run.py \
    --nuscenes \
    --future_frame_num 30 \
    --do_train \
    --data_dir "$train_dir" \
    --data_dir_for_val "$val_dir" \
    --output_dir "$output_dir" \
    --hidden_size 128 \
    --train_batch_size "$batch" \
    --use_map \
    --core_num 16 \
    --use_centerline \
    --distributed_training 1 \
    --other_params semantic_lane direction l1_loss goals_2D enhance_global_graph subdivide goal_scoring laneGCN point_sub_graph lane_scoring complete_traj complete_traj-3 \
    --eval_params optimization MRminFDE=0.0 cnt_sample=9 opti_time=0.1 \
    --learning_rate "$lr" \
    --hidden_dropout_prob "$dropout" \
    --argoverse \
    --argoverse2 \
    --num_train_epochs "$epochs"
```

#### evaluation
```bash
    python src/run.py \
      --nuscenes \
      --future_frame_num 30 \
      --do_eval \
      --data_dir "$train_dir" \
      --data_dir_for_val "$val_dir" \
      --output_dir "$output_dir" \
      --hidden_size 128 \
      --train_batch_size "$batch" \
      --use_map \
      --core_num 16 \
      --use_centerline \
      --distributed_training 1 \
      --other_params semantic_lane direction l1_loss goals_2D enhance_global_graph subdivide goal_scoring laneGCN point_sub_graph lane_scoring complete_traj complete_traj-3 \
      --eval_params optimization MRminFDE=0.0 cnt_sample=9 opti_time=0.1 \
      --learning_rate "$lr" \
      --hidden_dropout_prob "$dropout" \
      --argoverse \
      --argoverse2 \
      --model_recover_path "$model_path" >> "$output_dir/eval_results_$timestamp" \
      --visualize
```

My questions:
1. Do you see any (obvious) issues/mistakes in my training/evaluation commands?
2. During training, the model weights are saved after each epoch and the evaluation code is ran on each of these model weights. How do you choose the best model? Is it by choosing the one with best 'minADE' value?

**Comment by alfredgu001324** [comment]
Thanks for your comments!

Unfortunately this is indeed too long ago... I already forgot about the technical details. So I cannot really comment on these two questions :(. Really really sorry. 

But this might not be your problem, this part is actually done by my collaborator. But back then I also heard from him that DenseTNT performance is pretty unstable. If I remember correctly, you have a set of 16 checkpoints to choose from or whatever? And like the middle one is already the best iirc. 

Other people are also having problem replicating DenseTNT results, so you are not alone. It might just be that DenseTNT is quite sensitive to individual setups or whatever.. I always wonder whether there is a better way to encode the uncertainty info with GNN-based models but I left the field of trajectory prediction ever since then :(.

**Comment by alfredgu001324** [comment]
I think in my other repo, DenseTNT results also show that without uncertainty is better. I suspect it just depends on the hyper parameter config etc.. Every time we change something, we need to retune it. I am actually not sure whether this is actually a bug in my code... Because your number seems really bad.... sorry it's just been a while and could not remember clearly. 

I think in one of my issues, I might uploaded some raw code that are the un-clean version before I upload to GitHub. Maybe you can check on those to see whether there are any significant differences?

**Comment by rusty-electron** [comment]
Thanks for taking the time to reply. I really appreciate that you are active in the issues section. 

Is this the [issue](https://github.com/alfredgu001324/MapUncertaintyPrediction/issues/26) you mention?

**Comment by alfredgu001324** [comment]
Uhmmm I do not think so. But I somehow found it in this [issue](https://github.com/alfredgu001324/MapUncertaintyPrediction/issues/1) and traced it back to my email folder..... 

This should be the most raw version I have: https://drive.google.com/file/d/1kLfPWAVKAELBiEPeGwpFdlrMRDdTb3ct/view?usp=sharing 

If this still does not work then I guess DenseTNT is just not as stable as HiVT...

**Comment by rusty-electron** [comment]
Thanks for you efforts :),  I will go through the code.
